### PR TITLE
Add deps to repo

### PR DESCRIPTION
**Pull request by solidnerd**
This will make an build a lot of more comfortable. For dependency management dep will be used.

Otherwise every check out needs to have the correct dependencies and it's possible that it won't work in future anymore.

Signed-off-by: solidnerd <niclas@mietz.io>

### CODE REVIEW

**Comment by mxschmitt** [comment]
Hi solidnerd,

I'm glad that you want to contribute, Good Job!

@monochromata @JonasProgrammer What do you think about dependency management?
I see a few solutions to solve the kind of issue, that the driver could be broke by major changes in dependencies.

a) wait for the vgo release in go 1.11
b) use go dep, but don't include the vendor directory
c) use go dep, but include the vendor directory

Currently b) is the case for this pull request. But I think it's not so much necessary because we currently have only 3 dependencies and so we can wait for the official vgo release and make it "right" instead of bumping up this git repository by including the vendor directory.

Best regards

Max

**Comment by solidnerd** [comment]
Hey @mxschmitt,

thanks for your fast response.  My PR is option c) and not be, because I also include the vendor dir.

I think option a) is not a good option because it will be included in go  1.11 yes but it's in an [experimental state](https://github.com/golang/go/wiki/vgo#current-state) . So it will take some time at least 1 or more releases to get vgo stable. During the meantime the dependencies could brake and no one can build this cool project again ;) . 

dep is safe and production ready and will be used by a lot of people so we could go with b) or c) . 

Coming to option b)  this is also not the best choice because it forces new users to use dep to get the dependencies. If a new user wants to build this project he also needs dep for it to create the vendor directory. If we include this directly he can build the project directly and don't need any other tools than go since 1.5 the vendor directory will be used for deps.

**Comment by monochromata** [comment]
Hi, I'm not a go developer so I don't know what is common practice with that language. But I'd like to share my view.

My understanding is that few people will build the project anyways, since we provide binaries. When I built the project, the build itself did not cause issues I can recall. (I.e. I cannot see how the build could become "a lot of more comfortable".)

From my own practice I'd be ok if the build fails if the dependencies are gone or have changed in an incompatible way. That would mean that we need to update our code to work with the then-current versions of the dependencies. If dependencies are gone, we'd need to find new ones or abandon the project. I.e. I cater the thought that a project that is worth being used should be actively maintained.

Looking at our repo that contains 6 files with maybe a dozen dependencies if you count the import statements, it does not feel right to me to add 400+ files that merely cache stuff that is downloaded automatically.

So to me option a) sounds like a good idea.

Thanks for your initiative, though, Niclas!

**Comment by solidnerd** [comment]
@monochromata  If you got the problem with having about 400+ files.  We could add only `Gopkg.toml ` and `Gopkg.lock` to the project and write a little description for the development. 

> My understanding is that few people will build the project anyways, since we provide binaries. When I built the project, the build itself did not cause issues I can recall. (I.e. I cannot see how the build could become "a lot of more comfortable".)

It doesn't matter how many people build this. Another of using a dependency management tool is that we can ensure that we use the same dependencies during our local development and the build system. This will reduce the risk of bugs and unintended side effects.

**Comment by JonasProgrammer** [comment]
Hello everyone involved,

sorry to be the last to join the party. After reading through the PR discussion and skimming over `dep`'s documentation page, option `b)` is the most likable to me currently, which would basically be what @solidnerd suggested in his last comment as well.

As for option `a)`, while I only briefly had a look at `vgo`, the authors themselves currently recommend `dep` for production environments. And to my understanding even when shipped with golang 1.11, it will still be experimental; the main intention seems to be to have it tested by a broader audience.
Also, there is nothing that would prevent us from eventually migrating to `vgo` later on, should it be the preferred way of the `golang` community by then.

As for `c)` or the current state, the only real advantage I currently see is building being easier. Now, while this might broaden our potential audience a little, I think the amount is negligible. So far, I agree with @monochromata that people, who want it the 'easy' way will stick to binaries and those who want to hack on it can surely be persuaded into getting `dep` (maven, gradle, pip, gem, bundler, cmake, npm; one tool more or less installed will not worsen things now...).
As for the CI argument, we could add a plain and simple snippet in our `travis.yml` to `dep ensure` before building. Problem solved, without carrying 400 files.

The other problem I see would be pull requests with option `c)`. Someone might need to update some package some time, and with transitive dependencies being in play, I imagine this to escalate quite quickly. My point is: if somebody with malicious intend were to change something in the vendor directory, it might not be easy to spot. Sure, I could just checkout the PR, `dep ensure` and then diff afterwards to ensure (hehe) everything is in order, but this is some kind of step down from simply verifying PRs in the browser. Checking diffs of `Gopkg.toml` and and perhaps `Gopkg.lock` seems much more bearable.

Granted, the last example is somewhat constructed; there is way too few users for this to be a malware target. Still though, somebody touching dependencies is going to cause explosion in diffs.

Long story short, I am totally in for using `dep`. But I do not see the advantages of carrying around the vendor tree everywhere outweigh the disadvantages.

If I missed any points, open discussion is encouraged.

Best regards
JP

**Comment by solidnerd** [comment]
@monochromata @mxschmitt  @JonasProgrammer 

Okay I rebased my PR, that we can fix the dependencies. Also I updated the README to make it more clear how to use it. 

If anything is missing let me know othwise feel free to merge ;)

**Comment by JonasProgrammer** [comment]
LGTM

I've tweaked the travis script to honour the dependencies, as described [here](https://golang.github.io/dep/docs/FAQ.html#how-do-i-use-dep-in-ci).

If @monochromata and @mxschmitt agree, we can merge, I guess.

**Comment by monochromata** [comment]
Looks good to me.

**Comment by mxschmitt** [comment]
I've merged it.

@solidnerd Thanks!
